### PR TITLE
feat: "Spanify" DataProtector with IBufferWriter<byte>

### LINKED ISSUE #44758: API proposal: Add ReadOnlySpan<byte> to IDataProtector (un)Protect
## Background and Motivation

Add `ReadOnlySpan<byte>` overrides so consumer apps are able to avoid a extra array allocation+blockbuffer.copy every time they try to protect\unprotect data.

## Proposed API

```diff
namespace Microsoft.AspNetCore.DataProtection;

public interface IDataProtector : IDataProtectionProvider
{
+    byte[] Protect(ReadOnlySpan<byte> plaintext);

+    byte[] Unprotect(ReadOnlySpan<byte> protectedData);
}
```

## Usage Examples

```csharp
var buffer = new ArrayBufferWriter<byte>();

// write some data to buffer
var blob = new Data { Some = "Payload" };
MessagePackSerializer.Serialize(buffer, blob);

IDataProtector protector = provider.CreateProtector("demo");
return protector.Protect(buffer.WrittenSpan);
```

## Alternative Designs

```diff
namespace Microsoft.AspNetCore.DataProtection;

public interface IDataProtector : IDataProtectionProvider
{
+    byte[] Protect(ReadOnlySpan<byte> plaintext) => this.Protect(plaintext.ToArray())

+    byte[] Unprotect(ReadOnlySpan<byte> protectedData) => this.Unprotect(protectedData.ToArray())
}
```

## Risks

All of the existing protector implementations have to be changed to implement this change.
We could also have a default interface implementation to prevent it breaking to aggressively.

Performance impact should be minimal, and actually open the possibility for apps to avoid a unnecessary array copy, if people use the api correctly

### PR DESCRIPTION
**Pull request by DeagleGross**
_Please, review with caution! This is cryptography and failures here can be pretty dramatic..._

This is a 2nd iteration over spanification of DataProtection. [1st iteration used a different pattern](https://github.com/dotnet/aspnetcore/pull/62903): `GetSize()` followed by actual operation (protection or unprotection). Here we introduce `IBufferWriter<byte>` as destination parameter.

The following idea came up due to the internals of DataProtection: it may happen (though it's rare) that key changes during the user's interaction with DataProtection API, and suddenly the destination buffer can be of an insufficient size. Consider:
```
var protectBufferSize = dataProtector.GetSize();
var destination = ArrayPool<byte>.Shared.Rent(protectBufferSize);
// >>> IMAGINE key is changed under the hood, and now destination buffer will be twice bigger <<<
if (dataProtector.TryProtect(plainText, destination))
{
...
}
```
However, we can simply solve it by using `IBufferWriter<byte>` - since it has an ability to expand its size on-the-fly.

We have experimented with the API form a lot, but decided to go with such API. **Note that those APIs will be available only in NET11 or later**.
```diff
#if NET

+ public interface ISpanDataProtector : IDataProtector
{
+     void Protect<TWriter>(ReadOnlySpan<byte> plaintext, ref TWriter destination) 
+          where TWriter : IBufferWriter<byte>, allows ref struct;

+     void Unprotect<TWriter>(ReadOnlySpan<byte> protectedData, ref TWriter destination)
+          where TWriter : IBufferWriter<byte>, allows ref struct;
}

#endif
```

and 
```diff
#if NET

+public interface ISpanAuthenticatedEncryptor : IAuthenticatedEncryptor
{
+     void Encrypt<TWriter>(ReadOnlySpan<byte> plaintext, ReadOnlySpan<byte> additionalAuthenticatedData, ref TWriter destination)
+          where TWriter : IBufferWriter<byte>, allows ref struct;

+     void Decrypt<TWriter>(ReadOnlySpan<byte> ciphertext, ReadOnlySpan<byte> additionalAuthenticatedData, ref TWriter destination)
+          where TWriter : IBufferWriter<byte>, allows ref struct;
}

#endif
```

New API has a few benefits and intentions:
1) As described above, passing `IBufferWriter<byte>` avoids an undeterministic scenario where buffer may be not of enough size
2) Usage is simplier (no need for doing `GetSize()` then `Protect()/Unprotect()` with not 100% guarantee that works, which means a fallback to allocatey API is required).
3) It has a generic  parameter `TWriter` with a constraint to be `IBufferWriter<byte>` to allow passing in structs for even further optimization (reduction of allocation of a ref-type for buffer instance)
4) It has an anti-constraint and passes the buffer with `ref` to allows passing in a mutable ref struct. That should help in most of the cases DataProtection was designed to be used for - like Antiforgery where the input is quite small (under 100 bytes for instance), and buffer for encryption can be stackalloc'ed (which is probably the most performant mechanism to operate with a temporary buffer)

[Artificial benchmarks show expected results](https://github.com/DeagleGross/Baraholka/blob/dafa28e8dce09824c411c863b83d01e89e8dc9b8/Benchmarks/DIMvsDedicatedImplementation.cs#L24C1-L54C212): ref struct which uses stackalloc'ed buffer is the fastest way to operate.

That is also shown in the real benchmarks I've done on windows machine comparing original implementation, passing in a buffer which is a class, and a buffer whcih is a ref struct. With ref struct we get the most perf and use the least allocations.

|                                 Method |        Job |      Toolchain | RunStrategy | PlaintextLength |     Mean |     Error |    StdDev |   Median |      Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------------- |----------- |--------------- |------------ |---------------- |---------:|----------:|----------:|---------:|----------:|-------:|------:|------:|----------:|
|    ByteArray_ProtectUnprotectRoundtrip | Job-REJWKR | .NET Core 10.0 |  Throughput |              80 | 4.039 us | 0.0805 us | 0.2078 us | 3.939 us | 247,555.9 |      - |     - |     - |     512 B |
| PooledWriter_ProtectUnprotectRoundtrip | Job-REJWKR | .NET Core 10.0 |  Throughput |              80 | 3.809 us | 0.0751 us | 0.0835 us | 3.783 us | 262,504.4 |      - |     - |     - |     224 B |
|    RefWriter_ProtectUnprotectRoundtrip | Job-REJWKR | .NET Core 10.0 |  Throughput |              80 | 3.718 us | 0.0711 us | 0.0665 us | 3.717 us | 268,936.9 |      - |     - |     - |     160 B |

Fixes #44758

### CODE REVIEW

File: src/DataProtection/Abstractions/src/ISpanDataProtector.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /// <remarks>
+    /// This method provides an optimized, streaming alternative to <see cref="IDataProtector.Protect(byte[])"/>.
+    /// Rather than allocating an intermediate buffer, the protected data is written directly to the provided
+    /// buffer writer, which can improve performance and reduce memory allocation pressure.
+    /// The buffer writer is advanced by the total number of bytes written to it.
+    /// </remarks>
+    void Protect<TWriter>(ReadOnlySpan<byte> plaintext, ref TWriter destination)
+        where TWriter : IBufferWriter<byte>
+#if NET
+        , allows ref struct
```
Comment (halter73) [inline src/DataProtection/Abstractions/src/ISpanDataProtector.cs]: Would it be better to drop the netstandard support for this high-performance (perhaps unwieldy) API limit `ISpanDataProtector` and `ISpanAuthenticatedEncryptor` to .NET 11 and above?

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        if (!canAdvance)
+        {
+            throw new InvalidOperationException($"Cannot advance past the end of the buffer. Current position: {_index}, Capacity: {_buffer.Length}, Requested advance: {count}.");
+        }
+        _index += count;
+    }
+
+    private void CheckAndResizeBuffer(int sizeHint)
+    {
+        if (sizeHint <= 0)
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]:
```suggestion
        if (sizeHint < 0)
```

Good catch @Copilot. We should add a test case where `sizeHint` is zero.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+            var oldBuffer = _rentedBuffer;
+            _rentedBuffer = ArrayPool<T>.Shared.Rent(newSize);
+
+            Debug.Assert(oldBuffer.Length >= _index);
+            Debug.Assert(_rentedBuffer.Length >= _index);
+
+            var previousBuffer = oldBuffer.AsSpan(0, _index);
+            previousBuffer.CopyTo(_rentedBuffer);
+            previousBuffer.Clear();
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]:
```suggestion
```

Do we need to clear here? Presumably this will all be overwritten anyway. If we did want to clear, we should probably pass `clearArray: true` to `ArrayPool<byte>.Return` right below, but I don't think that's necessary.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        _index = -1;
+
+        if (_rentedBuffer is not null)
+        {
+            // Optionally clear the buffer before returning to pool (security)
+            // Uncomment if needed for sensitive data:
+            // _buffer.AsSpan(0, _index).Clear();
+
+            ArrayPool<T>.Shared.Return(_rentedBuffer, clearArray: false);
+            _buffer = null!;
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: No reason to leave the dead code around or to be explicit about `clearArray` when not clearing is the well-known default. But if we did want to reduce risk, we could create an ArrayPool using `ArrayPool<byte>.Create()`. that's used exclusively for DataProtection rather than use the shared one. This would also give us more control over the max size of pooled arrays and the maximum number of arrays to pool.

It's also unclear what nulling out the buffer does. Presumably the stack frame with the ref struct will be popped shortly after this will be disposed.

```suggestion
            ArrayPool<T>.Shared.Return(_rentedBuffer);
```

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        }
+        else
+        {
+            _buffer.Slice(0, _index).Clear();
+        }
+        _index = 0;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void ThrowIfDisposed()
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: Do we ever give user code access to this `IBufferWriter<T>`? If not, the most efficient thing would probably be to replace this with a `Debug.Assert`.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        }
+        Debug.Assert(_rentedBuffer.Length - _index > 0);
+        Debug.Assert(_rentedBuffer.Length - _index >= sizeHint);
+    }
+
+    /// <summary>
+    /// Clears the buffer, resetting the write position to zero without deallocating.
+    /// </summary>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void Clear()
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: What calls this? Even if we need this, I probably wouldn't bother doing more than resetting the index.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        {
+            var bufferSpace = _buffer.Length - _index;
+            if (bufferSpace < sizeHint)
+            {
+                // initial buffer is not enough, we need to start renting from the pool
+                var rentedInitialSize = _buffer.Length + Math.Max(sizeHint, _buffer.Length);
+                _rentedBuffer = ArrayPool<T>.Shared.Rent(rentedInitialSize);
+
+                _buffer.CopyTo(_rentedBuffer);
+                _buffer.Clear();
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: Wouldn't it make the rest of the code a lot simpler if we just reassigned `_buffer` here? At this point, we have no more use for the previous `_buffer`, right?

```suggestion
                _buffer = _rentedBuffer.AsSpan();
```

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+namespace System.Buffers;
+
+/// <summary>
+/// A high-performance struct-based IBufferWriter&lt;byte&gt; implementation that uses ArrayPool for allocations.
+/// Designed for zero-allocation scenarios when used with generic methods via `allows ref struct` constraint.
+/// </summary>
+internal ref struct RefPooledArrayBufferWriter<T> : IBufferWriter<T>, IDisposable
+{
+    private T[] _rentedBuffer;
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]:
```suggestion
    private T[]? _rentedBuffer;
```

This is often null, we shouldn't pretend that it isn't. I think this will be a lot more convenient once we point `_buffer` to `_rentedBuffer` after setting it to a non-null value.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+            {
+                return _rentedBuffer.AsMemory(0, _index);
+            }
+
+            // either we throw, or copy the Span<T> data into a heap-allocated array (via pooling), and then return Memory<T>.
+            // For demonstration, that it should not be used with Memory<T> unless done something specific, we throw here.
+            throw new InvalidOperationException("Cannot convert Span<T> to Memory<T>");
+        }
+    }
+
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: It appears there's really no safe way to call this, and it's not part of the IBufferWriter interface anyway. The best way to ensure it doesn't get called is to delete it.

```suggestion
```

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    /// Gets the number of bytes written to the buffer.
+    /// </summary>
+    public int WrittenCount
+    {
+        get
+        {
+            ThrowIfDisposed();
+            return _index;
+        }
+    }
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: This seems redundant with `WrittenSpan`. I suppose this is slightly faster, but it's only used for testing, so I'd go with less code.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+    /// <summary>
+    /// Gets the capacity of the underlying buffer.
+    /// </summary>
+    public int Capacity
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]:
```suggestion
    // For testing
    internal int Capacity
```

I know the whole type is internal, but I like calling out what's used by real code and what's just used for testing purposes. We should do the same for `FreeCapacity`.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+        CheckAndResizeBuffer(sizeHint);
+        if (_rentedBuffer is not null)
+        {
+            return _rentedBuffer.AsMemory(_index);
+        }
+
+        // either we throw, or copy the Span<T> data into a heap-allocated array (via pooling), and then return Memory<T>.
+        // For demonstration, that it should not be used with Memory<T> unless done something specific, we throw here.
+        throw new InvalidOperationException("Cannot convert Span<T> to Memory<T>");
+    }
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs:74]:
```suggestion
    public Memory<T> GetMemory(int sizeHint = 0)
    {
        throw new UnreachableException("RefPooledArrayBufferWriter does not support GetMemory");
    }
```

This is part of the interface, but it cannot be safely used. Since we're the only caller, we should just use an UnreachableException assuming we're not targeting netstandard with this. Trying to make it work when a buffer is rented just makes a heisenbug more likely. If we really did need to support this for some reason, we should just rent a buffer at this point if we haven't already.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        ThrowIfDisposed();
+        CheckAndResizeBuffer(sizeHint);
+        return _rentedBuffer is not null
+            ? _rentedBuffer.AsSpan(_index)
+            : _buffer.Slice(_index);
+    }
+
+    public void Advance(uint count)
+        => Advance((int)count);
+
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]:
```suggestion
```

Other than tests, don't we always have to call the IBufferWriter overload that takes int anyway? I see a bunch of calls to both `Advance(checked((int)` and just `Advance((int)`. We should probably be more consistent with it. Since we reject any count below zero, I think unchecked conversions from uint are fine.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    /// Initializes the <see cref="RefPooledArrayBufferWriter{T}"/> with initial buffer. 
+    /// </summary>
+    /// <param name="initialBuffer">The initial buffer to start writer with.</param>
+    public RefPooledArrayBufferWriter(Span<T> initialBuffer)
+    {
+        // no rented buffer initially - only if we need to grow over the limits of initialBuffer
+        _rentedBuffer = null!;
+
+        _buffer = initialBuffer;
+        _index = 0;
```
Comment (halter73) [inline src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: I don't think this comment is necessary once we properly mark the field as nullable.

```suggestion
        _buffer = initialBuffer;
```

File: src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs [hunk 15]
```diff
@@ -56,222 +57,162 @@ public CbcAuthenticatedEncryptor(Secret keyDerivationKey, BCryptAlgorithmHandle
...
+                        cbOutput: 0,
+                        pcbResult: out var dwRequiredSize,
+                        dwFlags: BCryptEncryptFlags.BCRYPT_BLOCK_PADDING);
+                    UnsafeNativeMethods.ThrowExceptionForBCryptStatus(ntstatus);
+
+                    // Get buffer from writer with the required size
+                    var buffer = destination.GetSpan(checked((int)dwRequiredSize));
+
+                    // Clone IV again for the actual decryption call
+                    byte* pbClonedIV2 = stackalloc byte[checked((int)_symmetricAlgorithmBlockSizeInBytes)];
```
Comment (halter73) [inline src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs]: This is the second call to `checked((int)_symmetricAlgorithmBlockSizeInBytes)`. While checked operations aren't *that* expensive, it's probably good to reuse the operation or just do the unchecked cast.

```suggestion
                    byte* pbClonedIV2 = stackalloc byte[(int)_symmetricAlgorithmBlockSizeInBytes];
```

File: src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs [hunk 16]
```diff
@@ -56,222 +57,162 @@ public CbcAuthenticatedEncryptor(Secret keyDerivationKey, BCryptAlgorithmHandle
...
+            {
+                ArrayPool<byte>.Shared.Return(rentedBuffer, clearArray: true);
+            }
         }
 
-        // Decryption finished!
-        CryptoUtil.Assert(dwActualDecryptedByteCount <= dwEstimatedDecryptedByteCount, "dwActualDecryptedByteCount <= dwEstimatedDecryptedByteCount");
-        if (dwActualDecryptedByteCount == dwEstimatedDecryptedByteCount)
+#else
+        var pooledArrayBuffer = new PooledArrayBufferWriter<byte>(outputSize);
```
Comment (halter73) [inline src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs:208]: Is there any reason we cannot use `using` for these?

```suggestion
        using var pooledArrayBuffer = new PooledArrayBufferWriter<byte>(outputSize);
```

File: src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs [hunk 17]
```diff
@@ -56,222 +57,162 @@ public CbcAuthenticatedEncryptor(Secret keyDerivationKey, BCryptAlgorithmHandle
...
+
+        var outputSize = ciphertext.Count - (int)(KEY_MODIFIER_SIZE_IN_BYTES + _symmetricAlgorithmBlockSizeInBytes + _hmacAlgorithmDigestLengthInBytes);
+        if (outputSize <= 0)
         {
             throw Error.CryptCommon_PayloadInvalid();
         }
 
-        // Assumption: pbCipherText := { keyModifier | IV | encryptedData | MAC(IV | encryptedPayload) }
+#if NET
+        byte[] rentedBuffer = null!;
```
Comment (halter73) [inline src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs]: As a general rule, we shouldn't use the null-forgiving operator when initializing variables/fields/properties, unless we're intentionally lying about the nullability of a public property that's always non-null by the time the user sees it, and even that's super questionable. It's better to save the null-forgiving until later in the code when we know for sure the variable/field/property must not be null. In this case, the only other place we use it we have a `if (rentedBuffer is not null)` anyways.

```suggestion
        byte[]? rentedBuffer = null;
```

This is really making me wish C# had macros though. I wonder if we should just have another overload to RefPooledArrayBufferWriter that takes an initial size other than a stackalloced Span. Not only would it simplify a lot of this repeated code a little bit, this would ensure the `RefPooledArrayBufferWriter` is always be responsible for renting and returning arrays to the pool.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 2 above)
Comment (halter73) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: I guess I never should have @-mentioned copilot. Of course, it opened a PR because of this 😆

File: src/DataProtection/Abstractions/src/ISpanDataProtector.cs (hunk 1 above)
Comment (DeagleGross) [reply src/DataProtection/Abstractions/src/ISpanDataProtector.cs]: ye, it makes sense. via #ifdef made it applicable only for NET (net11)

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 3 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: I agree, unless that is crypto specific and we need extra safety. I think we are good to not clear it.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 4 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: i removed it. I think runtime does use shared arrayPool, and no clearing is required there, so we should be ok as well (?)

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 6 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: nobody. I just tried to do the same API as exists in `class PooledArrayBufferWriter<T>`. I think we should be good to delete any unused API, and return if needed.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 5 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: agree, I thought that it is ok to show that to user, but hopefully our code with never throw

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 11 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: Honestly, we probabyl dont need that for tests as well

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 12 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs:74]: `UnreachableException` is not available in netstandard2.0, and `RefPooledArrayBufferWriter<T>` is in shared code, so I've implemented like this:
```csharp
#if NET
        throw new UnreachableException("RefPooledArrayBufferWriter does not support GetMemory");
#else
        throw new NotSupportedException("RefPooledArrayBufferWriter does not support GetMemory");
#endif
```

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 14 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: removed unnecessary code

File: src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs (hunk 16 above)
Comment (DeagleGross) [reply src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs:208]: We are getting CS1657: Cannot use `pooledArrayBuffer` as a ref or out value because it is a `using` variable.
image

That is yet another downside of using this API with non ref-struct as a `IBufferWriter<T>`, but it is what it is

File: src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs (hunk 17 above)
Comment (DeagleGross) [reply src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs]: 1) applied code suggestion to define `rentedBuffer` as `byte[]?`.
2) I think we should not add the public ctor to `RefPooledArrayBufferWriter` which would take the size for renting, because that encourages developers to use the stackalloc, and that is actually the goal for `RefPooledArrayBufferWriter`.

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 13 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: IAuthenticatedEncryptors have parameters `uint preBufferSize, uint postBufferSize`, so I wanted to have an overload, which simplifies the code instead of calling `Advance((int)preBufferSize)` for example.

but okay, removed it, calling it directly with casting

File: src/Shared/Buffers/RefPooledArrayBufferWriter.cs (hunk 7 above)
Comment (DeagleGross) [reply src/Shared/Buffers/RefPooledArrayBufferWriter.cs]: refactored as you proposed. not trivial for me, but makes sense after looking at the code

File: src/DataProtection/Abstractions/src/PublicAPI/net10.0/PublicAPI.Unshipped.txt [hunk 18]
```diff
@@ -0,0 +1,4 @@
+#nullable enable
```
Comment (halter73) [inline src/DataProtection/Abstractions/src/PublicAPI/net10.0/PublicAPI.Unshipped.txt:1]: Do we need to [update the API baselines](https://github.com/dotnet/aspnetcore/blob/13524176e22eaba6bc74d9d052ed91f629f6613b/docs/APIBaselines.md?plain=1)? @wtgodbe Is there any reason we should hold off on this?

Comment (wtgodbe) [reply src/DataProtection/Abstractions/src/PublicAPI/net10.0/PublicAPI.Unshipped.txt:1]: We should, yeah. No reason to hold off, I've already marked everything from 10 as shipped.

Comment (DeagleGross) [reply src/DataProtection/Abstractions/src/PublicAPI/net10.0/PublicAPI.Unshipped.txt:1]: I do not see any public changes since the [branching release/10.0](https://github.com/dotnet/aspnetcore/commit/896c176c5ab71e7ff7d7015ae4a7c2d0e486edc2).

I've ran the aspnetcore version of [mark-shipped.cmd](https://github.com/dotnet/aspnetcore/blob/main/eng/scripts/mark-shipped.cmd) today on main, and I do not see any files changed.

I assume all the work is already done? Build also passes for this PR, so current API is either added into Published or the new API in DataProtection is in the Unpublished files.

Let me know if I miss smth

Comment (DeagleGross) [reply src/DataProtection/Abstractions/src/PublicAPI/net10.0/PublicAPI.Unshipped.txt:1]: Ah, I guess I understand what is needed to be done.
A similar PR to https://github.com/dotnet/aspnetcore/pull/58090/commits/87cf9b959961f43c5635e8d8014ae7278f3b8f3b

Comment (DeagleGross) [reply src/DataProtection/Abstractions/src/PublicAPI/net10.0/PublicAPI.Unshipped.txt:1]: will be done here: https://github.com/dotnet/aspnetcore/pull/64627

Comment (wtgodbe) [reply src/DataProtection/Abstractions/src/PublicAPI/net10.0/PublicAPI.Unshipped.txt:1]: It's fine to leave the file named `net10.0` until we update the TFM - no need to block this PR on https://github.com/dotnet/aspnetcore/pull/64627. I already marked the 10.0 API as shipped in https://github.com/dotnet/aspnetcore/pull/64319

**Review by halter73 — APPROVED** [review]
I'm excited to see ant forgery take advantage of this! Thanks for your considerable effort evaluating all the different approaches.
